descwl_shear_sims/render_sim.py

```diff
@@ -114,4 +126,17 @@ def render_objs_with_psf_shear(
         if oshape[0]*oshape[1] > 0:
             se_im[overlap] += stamp[overlap]
 
-    return se_im
+            if (obj_data['type'] == 'star' and obj_data['saturated']):
```

Comment:
I'm not a huge fan of having this done here in this function. The only dependence needed is the star position. It seems like we should be able to abstract this out.

Reply:
do you mean factor this out?  What do you suggest?

Reply:
The bleeds should be part of rendering, since they are rendering

Do you no like setting a mask here?

Reply:
this function is for render galaxies etc. - it setting mask bits is weird to me and that will make it harder to test - we should have a separate function for that and we should test it properly

Reply:
The argument against that is it will involve duplicating a bunch of code that is within that loop.

To avoid that I would want to refactor all that stuff into a separate function

Reply:
As it is currently written, we would have to re-render the stamps, or save them for later, in order to do the overlap calculation

I guess one way around that is to render the bleed in the case that the center is located in the image rather than just when there is overlap, but that would mean that saturation would only occur in some of the epochs

Reply:
To put it more clearly:  that would mean that there are cases where there should be saturation but there isn't

Reply:
a better option would be to save a boolean whether there was overlap

Reply:
Im confused by these comments. The code doing the wcs transforms in the loop is general enough that we could abstract it. However I don’t want this to be a sticking point. My main concern is testing. I wrote a bunch of tests of this function and now with an extra branch we will need more of them.

Reply:
It also involves refactoring the entire code base for bright star masking, because I combine the adding of saturation and mask setting there

Reply:
Yeah that refactor should be done anyways to be honest. We are spraying and mangling state all over the place at some level.

Reply:
Look at this function:

https://github.com/LSSTDESC/descwl-shear-sims/blob/master/descwl_shear_sims/gen_star_masks.py#L224

Why would we write two separate functions, each of which do this same book keeping and calculations, with the only difference being in the last line?

Reply:
I'm going to take this stuff out of the render function because I think its fine, but I don't agree that we should duplicate code like the above

Reply:
Right. We should abstract the wcs handling too.

Reply:
If we are going to store a dict of info for each object, we can double down on that and store the wcs image positions and Jacobians there. Then each render function can consume that info.

Reply:
How about the recent commits?

Reply:
Right this is closer. My thought is to take the WCS code and stamp computation code here: https://github.com/LSSTDESC/descwl-shear-sims/pull/66/files#diff-5873b4ec62708400cfd519b1e39dfc5bR87 and move it to a separate function. 

Then we run this function once per object and epoch combination. The results we need get stored in the dict attached to each object. Then we feed that data back into a separate function that renders the object and another function the builds the masks. 

This separates out a lot of the code and keeps the code paths pretty linear and independent.

Reply:
I know I am pushing hard on the code structure here. My intent is that this code will live with us for a long time, so I want to get it right now.

Reply:
We still cannot do the overlap calculation until we render the stamp

Reply:
So we do a prerender of the stamp that is really fast to get the stamp size. This doesn't actually render the object. It is this part we should move out.

Reply:
From that and the extra info, we can compute the stamp bounds and overlap.

Reply:
We don't want to add them to the object dict which is per band not per epoch, unless we restructure that as well

Reply:
Well, we can store anything there, so as long as we are keeping extra state around, we might as well codify that into a formal API that we use consistently.

Reply:
I'm looking at this code and it just isn't obvious to me how to do what you are asking.  It is your code not mine so it isn't so easy to envision it.

I propose we merge this and open an new issue about refactoring, which you can work on when you get back from your leave. I doubt I will be implementing any new features that will require rendering beyond what we currently have:  at the most I plan to improve the drawing of saturated regions around stars, and improve bleed shapes, which fits in fine with the current setup.

Reply:
To be clear, I'm disagreeing with you, I just don't want to attempt it myself

Reply:
sorry, I'm *not* disagreeing :)

Reply:
Works for me!